pkg/cluster/loadbalancerprofile.go

```diff
@@ -0,0 +1,396 @@
+		Name:     &name,
+		ID:       &resourceID,
+		Location: &m.doc.OpenShiftCluster.Location,
+		PublicIPAddressPropertiesFormat: &mgmtnetwork.PublicIPAddressPropertiesFormat{
+			PublicIPAllocationMethod: mgmtnetwork.Static,
+			PublicIPAddressVersion:   mgmtnetwork.IPv4,
+		},
+		Sku: &mgmtnetwork.PublicIPAddressSku{
+			Name: mgmtnetwork.PublicIPAddressSkuNameStandard,
+		},
+	}
+
+	err := m.publicIPAddresses.CreateOrUpdateAndWait(ctx, resourceGroupName, name, publicIPAddress)
+	if err != nil {
+		return mgmtnetwork.PublicIPAddress{}, err
+	}
+
+	return publicIPAddress, nil
+}
+
+func getOutboundIPsFromLB(lb mgmtnetwork.LoadBalancer) []api.ResourceReference {
+	var outboundIPs []api.ResourceReference
+	fipConfigs := make(map[string]mgmtnetwork.FrontendIPConfiguration, len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations))
+
+	for i := 0; i < len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations); i++ {
+		fipConfigID := *(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations)[i].ID
+		fipConfig := (*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations)[i]
+		fipConfigs[fipConfigID] = fipConfig
+	}
+
+	for _, obRule := range *lb.LoadBalancerPropertiesFormat.OutboundRules {
+		if *obRule.Name == outboundRuleV4 {
+			for i := 0; i < len(*obRule.OutboundRulePropertiesFormat.FrontendIPConfigurations); i++ {
+				id := *(*obRule.OutboundRulePropertiesFormat.FrontendIPConfigurations)[i].ID
+				if fipConfig, ok := fipConfigs[id]; ok {
+					outboundIPs = append(outboundIPs, api.ResourceReference{ID: *fipConfig.PublicIPAddress.ID})
+				}
+			}
+		}
+	}
+
+	return outboundIPs
+}
+
+func (m *manager) patchEffectiveOutboundIPs(ctx context.Context, outboundIPs []api.ResourceReference) error {
+	m.log.Info("patching effectiveOutboundIPs")
+	effectiveOutboundIPs := make([]api.EffectiveOutboundIP, 0, len(outboundIPs))
+	for _, obIP := range outboundIPs {
+		effectiveOutboundIPs = append(effectiveOutboundIPs, api.EffectiveOutboundIP(obIP))
+	}
+	var err error
+	m.doc, err = m.db.PatchWithLease(ctx, m.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
+		doc.OpenShiftCluster.Properties.NetworkProfile.LoadBalancerProfile.EffectiveOutboundIPs = effectiveOutboundIPs
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func newFrontendIPConfig(name string, id string, publicIPorIPPrefixID string) mgmtnetwork.FrontendIPConfiguration {
+	// TODO: add check for publicIPorIPPrefixID
+	return mgmtnetwork.FrontendIPConfiguration{
+		Name: &name,
+		ID:   &id,
+		FrontendIPConfigurationPropertiesFormat: &mgmtnetwork.FrontendIPConfigurationPropertiesFormat{
+			PublicIPAddress: &mgmtnetwork.PublicIPAddress{
+				ID: &publicIPorIPPrefixID,
+			},
+		},
+	}
+}
+
+func newOutboundRuleFrontendIPConfig(id string) mgmtnetwork.SubResource {
+	return mgmtnetwork.SubResource{
+		ID: &id,
+	}
+}
+
+// Reports if two []api.ResourceReference are equal.
+func areResourceRefsEqual(a, b []api.ResourceReference) bool {
```

Comment:
Could we use `reflect.DeepEqual` here?

Reply:
I did change this function to return the result of `reflect.DeepEqual` but I found the lists must be sorted and lowercased for reflect.DeepEqual to work correctly.